### PR TITLE
potential version bump

### PR DESCRIPTION
**Pull request by charlesmulder**
…okens

removed a given from scenario

started initial work

Update authorization.feature

stub passing for authorizing of consumption

stub now returns token for consuming of configs

added some docs

use identifier to authorize consumption

authorizing consumption now only supported by admin client

was missing an or

added delegation feature

added delegation scenario to authorization feature

improved delegation feature and added scenario for creating credentials to authorization feature

saving progress

refactored features and steps

happy with generating of credentials to consume configuration

factory now making use of client

fixed unit tests

removed unit test fixes in favour of major version bump

updated scenario to ensure authorized consumption tokens can't publish

added required given to cs failure scenarios

added rdocs and removed redundant code

### CODE REVIEW

File: features/step_definitions/delegation_steps.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Given(/^I am not allowed to delegate access to consume my configuration$/) do
```
Comment (charlesmulder) [inline features/step_definitions/delegation_steps.rb]: this file shouldn't be here anymore

File: lib/configuration_service/admin_client.rb [hunk 2]
```diff
@@ -31,12 +31,19 @@ def request_configuration(identifier)
...
+    
+    ##
+    # Authorize consumption of a configuration ie. get credentials to consume configurations ie. get a token to consume configurations
+    # 
+    # @param [String] configuration identifier
+    # @return [String|Vault::Secret] token either directly available or at Vault::Secret.auth.client_token
+    # @raise [ConfigurationService::ConfigurationNotFoundError] if no configuration was found for the configured +identifier+
+    # @raise [ConfigurationService::Error] if the request failed
+    ##
+    def authorize_consumption(identifier)
```
Comment (sheldonh) [inline lib/configuration_service/admin_client.rb]: Why are we extending this class while deprecating it?

File: lib/configuration_service/client.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+require "configuration_service/utils"
+
+module ConfigurationService
+  class Client
+
+    ##
```
Comment (sheldonh) [inline lib/configuration_service/client.rb]: Documentation got worse (intent lost)

File: lib/configuration_service/client.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    # Authorize consumption of a configuration ie. get credentials to consume configurations ie. get a token to consume configurations
+    # 
+    # @param [String] unique identifier of configuration
+    # @return [String|Vault::Secret] credentials allowing consumption of the provided configuration
+    # @raise [ConfigurationService::ConfigurationNotFoundError] if no configuration was found for the configured +identifier+
+    # @raise [ConfigurationService::Error] if the request failed
+    ##
+    def authorize_consumption(identifier)
+      @provider.authorize_consumption(identifier, @credentials) or
+        raise ConfigurationNotFoundError, "configuration not found for identifier: #{@identifier}"
```
Comment (sheldonh) [inline lib/configuration_service/client.rb]: `@identifier` should be `identifier` here.

File: lib/configuration_service/provider/stub.rb [hunk 5]
```diff
@@ -58,24 +58,37 @@ def request_configuration(identifier, token)
       ##
       # Publish configuration
       #
-      # @param [ConfigurationService::Configuration] configuration configuration to publish
+      # @param [ConfigurationService::Configuration] configuration to publish
```
Comment (sheldonh) [inline lib/configuration_service/provider/stub.rb]: Missing param.

File: lib/configuration_service/provider/stub.rb [hunk 6]
```diff
@@ -47,8 +47,8 @@ def initialize(options = {})
       #
       # @see ConfigurationService::Base#request_configuration
       #
-      def request_configuration(identifier, token)
-        authorize_request(:consumer, token)
+      def request_configuration(identifier, credentials)
```
Comment (sheldonh) [inline lib/configuration_service/provider/stub.rb]: Document params.

File: lib/configuration_service/provider/stub.rb [hunk 7]
```diff
@@ -58,24 +58,37 @@ def request_configuration(identifier, token)
...
+      def publish_configuration(configuration, credentials)
+        authorize_request(:publisher, credentials)
         @configurations.store(configuration.identifier, configuration.data, configuration.metadata)
         configuration
       end
 
+      ##
+      # Authorizes consumption of configuration
+      #
+      # @param [Sting] identifier of configuration to be consumed
```
Comment (sheldonh) [inline lib/configuration_service/provider/stub.rb]: `Sting` should be `String`

File: lib/configuration_service/test/orchestration_provider.rb [hunk 8]
```diff
@@ -93,35 +96,35 @@ def delete_configuration
       end
 
       ##
-      # Provide a token that authorizes a role
+      # Provide a credentials that authorizes a role
```
Comment (sheldonh) [inline lib/configuration_service/test/orchestration_provider.rb]: Provide credentials. :-)

File: lib/configuration_service/test/stub_orchestration_provider.rb [hunk 9]
```diff
@@ -51,13 +51,13 @@ def broken_service_provider
       end
 
       ##
-      # Provide a token that authorizes a role
+      # Provide a credentials that authorizes a role
       #
-      # The token is taken from {ConfigurationService::Provider::Stub::BUILTIN_TOKENS}
+      # The credentials is taken from {ConfigurationService::Provider::Stub::BUILTIN_TOKENS}
```
Comment (sheldonh) [inline lib/configuration_service/test/stub_orchestration_provider.rb]: The credentials are...

**Comment by sheldonh** [comment]
Notes added, otherwise very happy.

**Pull request merged**